src/prune.ts

```diff
@@ -0,0 +1,186 @@
+  devDependencies?: { [type: string]: string }
+}
+
+function readMasterDependencies(projectDirectory: string): Promise<TypeDependencies> {
+  return readConfig(join(projectDirectory, CONFIG_FILE))
+}
+
+interface PruneExtraneousOptions {
+  masterDependencies: TypeDependencies
+  typingsPath: string
+  includeDev: boolean
+  dtsPath: string
+  emitter: Emitter
+}
+
+function pruneExtraneous(options: PruneExtraneousOptions): Promise<void> {
+  const { masterDependencies, typingsPath, includeDev, dtsPath, emitter } = options
+
+  return transformDtsFile(dtsPath, (typings: string[]): Promise<string[]> => {
+    // Reverse parse reference paths into dependency info
+    const dependencies = typings.map(typing => parseReferencePath(typingsPath, typing))
+
+    // Separate dependencies into extraneous and used
+    const [extraneous, used] = partition(dependencies, dependency => isExtraneous(masterDependencies, dependency, includeDev))
+
+    // Parse extraneous
+    return Promise.all(
+      extraneous.map(dependency => pruneTypings({
+        typingsDirectory: dependency.directory,
+        typingsFile: dependency.path,
+        typingsName: dependency.name,
+        emitter
+      }))
+    )
+      .then(() => {
+        // Return used for dts
+        return used.map(dependency => dependency.path)
+      })
+  })
+}
+
+interface DependencyInfo {
+  path: string
+  directory: string
+  name: string
+  isAmbient: boolean
+}
+
+function parseReferencePath(typingsDirectory: string, path: string): DependencyInfo {
+  const relativeTypingsPath = relativeTo(typingsDirectory, path)
+  const parts = relativeTypingsPath.split(pathSeparator)
+
+  // Remove 'typings' folder
+  parts.shift()
+
+  // Remove .d.ts file
+  if (isDefinition(path)) {
+    parts.pop()
+  }
+
+  const [ /* main|browser */ , type, ...names] = parts
+  const relativeDirPath = parts.join(pathSeparator)
+
+  return {
+    path,
+    directory: join(typingsDirectory, relativeDirPath),
+    name: names.join(pathSeparator),
+    isAmbient: type === AMBIENT_TYPE_DIR
+  }
+}
+
+function isExtraneous(masterDependencies: TypeDependencies, dependency: DependencyInfo, includeDev: boolean): boolean {
+  if (dependency.isAmbient) {
+    if (masterDependencies.ambientDependencies &&
+        masterDependencies.ambientDependencies[dependency.name]) {
+      return false
+    } else if (includeDev &&
+               masterDependencies.ambientDevDependencies &&
+               masterDependencies.ambientDevDependencies[dependency.name]) {
+      return false
+    }
+  } else {
+    if (masterDependencies.dependencies &&
+        masterDependencies.dependencies[dependency.name]) {
+      return false
+    } else if (includeDev &&
+               masterDependencies.devDependencies &&
+               masterDependencies.devDependencies[dependency.name]) {
+      return false
+    }
+  }
+  return true
+}
+
+function partition<T>(array: T[], predicate: { (item: T): boolean }): [T[], T[]] {
```

Comment:
This is a general purpose utility. There are some possibilities for changing this if desired:
* Move to utils
* Replace with an external dep from npm
* Replace with consecutive `array.prototype.filter()` calls on the predicate inverted.